{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "79a62ab5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import functions from these libraries\n",
    "\n",
    "import seaborn as sns\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import scipy.stats as stats\n",
    "from scipy.stats import norm\n",
    "import statistics\n",
    "import math\n",
    "from statsmodels.stats.weightstats import ztest as ztest\n",
    "sns.set_theme()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "36f49871",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta_file(input_file):\n",
    "    \"\"\"This parses FASTA files and compiles them into a dictionary \n",
    "    with the identity tag as the keys\n",
    "    and the nucleotide or amino acid sequences as the values\n",
    "    \"\"\"\n",
    "    \n",
    "    global parsed_seqs\n",
    "    parsed_seqs = {}\n",
    "\n",
    "    f = open(input_file)\n",
    "\n",
    "    curr_seq_id = None\n",
    "    curr_seq = []\n",
    "\n",
    "    for line in f:\n",
    "\n",
    "            line = line.strip()\n",
    "\n",
    "            if line.startswith(\">\"):\n",
    "                if curr_seq_id is not None:\n",
    "                    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "\n",
    "                curr_seq_id = line[1:]\n",
    "                curr_seq = []\n",
    "                continue\n",
    "\n",
    "            curr_seq.append(line)\n",
    "\n",
    "    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "    \n",
    "    print('Parse Complete')\n",
    "    print('Dictionary name is: parsed_seqs')\n",
    "    \n",
    "    global reference_protein\n",
    "    global reference_sequence\n",
    "    \n",
    "    reference_protein = list(parsed_seqs)[0]\n",
    "    reference_sequence = list(parsed_seqs.values())[0]\n",
    "    \n",
    "    print('Reference sequence variable name is: reference_sequence')\n",
    "    # separately, retrieve the first protein sequence in the dictionary\n",
    "    # and display it as a reference protein for subsequence location\n",
    "    \n",
    "    return f'Reference Protein: {reference_protein}, {reference_sequence}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f923065a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_sequence_identity(seq1,seq2):\n",
    "    \"\"\"Return the level of sequence identity between seq1 and seq2\"\"\"\n",
    "\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(f\"Aligned Sequences should be of equal length. Sequence 1 is {len(seq1)} nucleotides long but Sequence 2 is {len(seq2)} is nucleotides long.\")\n",
    "    \n",
    "    matches = 0\n",
    "    \n",
    "    for i,residue in enumerate(seq1):\n",
    "\n",
    "        if (i + 1) > len(seq2):\n",
    "            continue\n",
    "\n",
    "        residue2 = seq2[i]\n",
    "\n",
    "        if residue.upper() == residue2.upper():\n",
    "            matches += 1\n",
    "\n",
    "    sequence_length = len(seq1)\n",
    "    sequence_identity = matches/sequence_length\n",
    "    \n",
    "    print(sequence_identity)\n",
    "    \n",
    "    return sequence_identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "2859b48d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def calculate_subsequence_identity(seq1,seq2,loc1,loc2):\n",
    "    \"\"\"Return the level of subsequence identity between seq1 and seq2 in a frame bounded by loc1 and loc2\"\"\"\n",
    "\n",
    "    if len(seq1) != len(seq2):\n",
    "        raise ValueError(f\"Aligned Subsequences should be of equal length. Sequence 1 is {len(seq1)} nucleotides long but Sequence 2 is {len(seq2)} is nucleotides long.\")\n",
    "        \n",
    "    if loc1 == loc2:\n",
    "        raise ValueError(\"Subsequence locations must be at least 1 nucleotide apart.\")\n",
    "        \n",
    "    matches = 0\n",
    "    \n",
    "    subseq1 = seq1[loc1 - 1:loc2 - 1]\n",
    "    subseq2 = seq2[loc1 - 1:loc2 - 1]\n",
    "    # location numbers are converted from positive integers to their index equivalents\n",
    "    # (e.g. the first residue of a subsequence is position 1)\n",
    "    \n",
    "    for i,residue in enumerate(subseq1):\n",
    "\n",
    "        if (i + 1) > len(subseq2):\n",
    "            continue\n",
    "\n",
    "        residue2 = subseq2[i]\n",
    "\n",
    "        if residue.upper() == residue2.upper():\n",
    "            matches += 1\n",
    "\n",
    "    subsequence_length = len(subseq1)\n",
    "    subsequence_identity = matches/subsequence_length\n",
    "\n",
    "    print(subsequence_identity)\n",
    "    \n",
    "    return subsequence_identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "498e24ba",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_multiple_sequence_identities(parsed_seqs,loc1,loc2):\n",
    "    \"\"\"Find the sequence identity for every pairwise comparison of nucleotide or amino acid sequences\n",
    "    and the identity for a subsequence beginning at loc1 and ending at loc2\n",
    "    of the sequences in a fasta file\n",
    "    then compile both into dictionaries with the comparison as the keys and the identity value as the values.\"\"\"\n",
    "    \n",
    "    global sequence_identities\n",
    "    global subsequence_identities\n",
    "    \n",
    "    sequence_identities = {}\n",
    "    subsequence_identities = {}\n",
    "    \n",
    "    # open new dictionaries for the sequence and subsequence identities\n",
    "    \n",
    "    print(\"Calculating\")\n",
    "    \n",
    "    location1 = loc1\n",
    "    location2 = loc2\n",
    "    \n",
    "    sequences = parsed_seqs\n",
    "\n",
    "    for identity_tag, sequence1 in sequences.items():\n",
    "        # iterate over every sequence in your fasta file\n",
    "\n",
    "        for identity_tag2, sequence2 in sequences.items():\n",
    "            # compare with every sequence in your fasta file\n",
    "\n",
    "            print(f\"> Comparing {identity_tag} with {identity_tag2}\")\n",
    "            \n",
    "            sequence_identity = calculate_sequence_identity(sequence1,sequence2)\n",
    "            subsequence_identity = calculate_subsequence_identity(sequence1,sequence2,location1,location2)\n",
    "            # return sequence and subsequence identities\n",
    "\n",
    "            comparison = f\"{sequence1}_{sequence2}\"\n",
    "\n",
    "            sequence_identities[comparison] = sequence_identity\n",
    "            subsequence_identities[comparison] = subsequence_identity\n",
    "            # compile into sequence and subsequence identity dictionaries\n",
    "    \n",
    "    print(\"Done\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6826e349",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_all(sequence,subsequence):\n",
    "    \"\"\"Find a predetermined subsequence (substring) in a longer sequence (string) including overlaps.\"\"\"\n",
    "    \n",
    "    start = 0\n",
    "\n",
    "    while True:\n",
    "        start = sequence.find(subsequence,start)\n",
    "        if start == -1: return\n",
    "        yield start + 1 \n",
    "        start += 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "ae928df5",
   "metadata": {},
   "outputs": [],
   "source": [
    "def find_subsequence(sequence,subsequence):\n",
    "    \"\"\"Display the subsequence locations in a list\n",
    "    and provide the subsequence length to be added\n",
    "    to find the optimal reading frame.\"\"\"\n",
    "    \n",
    "    subsequence_locations = list(find_all(sequence,subsequence))\n",
    "    return f\"Subsequence start locations: {subsequence_locations}\", f\"+{len(subsequence)}\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "ac03a671",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sequence_comparison(sequence_identities,subsequence_identities):\n",
    "\n",
    "    seq_id_list = list(sequence_identities.values())\n",
    "    subseq_id_list = list(subsequence_identities.values())\n",
    "    # move the sequence and subsequence identities from the dictionaries into lists\n",
    "\n",
    "    mean = statistics.mean(seq_id_list)\n",
    "    sd = statistics.stdev(seq_id_list)\n",
    "    mean2 = statistics.mean(subseq_id_list)\n",
    "    sd2 = statistics.stdev(subseq_id_list)\n",
    "\n",
    "    print(\"Sequence identity mean:\", mean)\n",
    "    print(\"Sequence identity standard deviation:\", sd)\n",
    "    print(\"Subsequence identity mean:\", mean2)\n",
    "    print(\"Subsequence identity standard deviation:\", sd2)\n",
    "    # calculate the mean and standard deviation of the identities, then display\n",
    "\n",
    "    if mean < mean2:\n",
    "        print('Subsequence is more conserved than sequence')\n",
    "    if mean > mean2:\n",
    "        print('Subsequence is less conserved than sequence')\n",
    "    if mean == mean2:\n",
    "        print('Subsequence is as conserved as sequence')\n",
    "    # clarify relationship of subsequence to sequence\n",
    "\n",
    "    sns.displot(seq_id_list, x=seq_id_list, kind=\"kde\", bw_adjust=1)\n",
    "    plt.xticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "    plt.xlabel('Sequence Identity')\n",
    "    # plot distribution of sequence identities\n",
    "    \n",
    "    sns.displot(subseq_id_list, x=subseq_id_list, kind=\"kde\", bw_adjust=2)\n",
    "    plt.xticks([0, 0.2, 0.4, 0.6, 0.8, 1.0])\n",
    "    plt.xlabel('Subsequence Identity')\n",
    "    # plot distribution of subsequence identities\n",
    "    \n",
    "    return stats.ttest_rel(subseq_id_list, seq_id_list), stats.ttest_1samp(subseq_id_list, mean)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "bdd812f4",
   "metadata": {},
   "outputs": [
    {
     "ename": "FileNotFoundError",
     "evalue": "[Errno 2] No such file or directory: ''",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mFileNotFoundError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[1;32mIn [9]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[43mparse_fasta_file\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[1;32mIn [2]\u001b[0m, in \u001b[0;36mparse_fasta_file\u001b[1;34m(input_file)\u001b[0m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;28;01mglobal\u001b[39;00m parsed_seqs\n\u001b[0;32m      8\u001b[0m parsed_seqs \u001b[38;5;241m=\u001b[39m {}\n\u001b[1;32m---> 10\u001b[0m f \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mopen\u001b[39;49m\u001b[43m(\u001b[49m\u001b[43minput_file\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m     12\u001b[0m curr_seq_id \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;01mNone\u001b[39;00m\n\u001b[0;32m     13\u001b[0m curr_seq \u001b[38;5;241m=\u001b[39m []\n",
      "\u001b[1;31mFileNotFoundError\u001b[0m: [Errno 2] No such file or directory: ''"
     ]
    }
   ],
   "source": [
    "parse_fasta_file('')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "6ce92a44",
   "metadata": {},
   "outputs": [],
   "source": [
    "find_subsequence(reference_sequence,'')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9103df1",
   "metadata": {},
   "outputs": [],
   "source": [
    "analyze_multiple_sequence_identities(parsed_seqs,,)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ad8f71e",
   "metadata": {
    "scrolled": false
   },
   "outputs": [],
   "source": [
    "sequence_comparison(sequence_identities,subsequence_identities)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d5413033",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
