{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "1a813897",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta_file(input_file):\n",
    "    \"\"\"This parses FASTA files and compiles them into a dictionary \n",
    "    with the identity tag as the keys\n",
    "    and the nucleotide or amino acid sequences as the values\n",
    "    \"\"\"\n",
    "    \n",
    "    global parsed_seqs\n",
    "    parsed_seqs = {}\n",
    "\n",
    "    f = open(input_file)\n",
    "\n",
    "    curr_seq_id = None\n",
    "    curr_seq = []\n",
    "\n",
    "    for line in f:\n",
    "\n",
    "            line = line.strip()\n",
    "\n",
    "            if line.startswith(\">\"):\n",
    "                if curr_seq_id is not None:\n",
    "                    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "\n",
    "                curr_seq_id = line[1:]\n",
    "                curr_seq = []\n",
    "                continue\n",
    "\n",
    "            curr_seq.append(line)\n",
    "\n",
    "    parsed_seqs[curr_seq_id] = ''.join(curr_seq)\n",
    "    \n",
    "    print('Parse Complete')\n",
    "    print('Dictionary name is: parsed_seqs')\n",
    "    \n",
    "    global reference_protein\n",
    "    global reference_sequence\n",
    "    \n",
    "    reference_protein = list(parsed_seqs)[0]\n",
    "    reference_sequence = list(parsed_seqs.values())[0]\n",
    "    \n",
    "    print('Reference sequence variable name is: reference_sequence')\n",
    "    # separately, retrieve the first protein sequence in the dictionary\n",
    "    # and display it as a reference protein for subsequence location\n",
    "    \n",
    "    return f'Reference Protein: {reference_protein}, {reference_sequence}'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "bff26f89",
   "metadata": {},
   "outputs": [],
   "source": [
    "def analyze_nucleotide_abundance(parsed_seqs):\n",
    "\n",
    "    global genome_string\n",
    "    genome_string = \"\"\n",
    "    #creates an empty string labeled \"genome_string\"\n",
    "    \n",
    "    sequences = parsed_seqs\n",
    "    \n",
    "    for identity_tag, sequence in sequences.items():\n",
    "        \n",
    "        genome_string += sequence\n",
    "        # compiles all residues from an annotated genome into a single string labeled genome_string\n",
    "    \n",
    "    genome_string = genome_string.upper()\n",
    "    # converts all nucleotide letters in genome_string to upper case\n",
    "    \n",
    "    total_count = len(genome_string)\n",
    "    print(\"Total Length: \",total_count)\n",
    "        \n",
    "    g_count = genome_string.count(\"G\")\n",
    "    g_percent = round(g_count / total_count, 4) * 100\n",
    "    print(f\"Number of G's: \",g_count,\",\",g_percent,\"%\")\n",
    "    \n",
    "    c_count = genome_string.count(\"C\")\n",
    "    c_percent = round(c_count / total_count, 4) * 100\n",
    "    print(f\"Number of C's: \",c_count,\",\",c_percent,\"%\")\n",
    "    \n",
    "    a_count = genome_string.count(\"A\")\n",
    "    a_percent = round(a_count / total_count, 4) * 100\n",
    "    print(f\"Number of A's: \",a_count,\",\",a_percent,\"%\")\n",
    "    \n",
    "    t_count = genome_string.count(\"T\")\n",
    "    t_percent = round(t_count / total_count, 4) * 100\n",
    "    print(f\"Number of T's: \",t_count,\",\",t_percent,\"%\")\n",
    "    \n",
    "    cg_count = genome_string.count(\"CG\")\n",
    "    cg_percent = round(cg_count / total_count, 4) * 100\n",
    "    print(f\"Number of CG's: \",cg_count,\",\",cg_percent,\"%\")\n",
    "    \n",
    "    # returns and prints the number and proportion of mononucleotides as well as CG dinucleotides\n",
    "    \n",
    "    return total_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ff9a80c5",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
